### ISSUE TITLE
Feature request: allow for NLog to render nested innerexceptions to JSON

### ISSUE THREAD

**Issue by mccabela** [NLog/NLog#1311]
Feature request for NLog version: 4.2.3

The current JSONLayout renderer only supports the rendering of the first exception in fully structured JSON.  I would like to be able to render out all the exceptions, including the inner exceptions, in structured JSON format.

This would allow for the separate exceptions to be inspected and allow for a deeper understanding of what happened.

Currently you can do something like this:

``` xml
<layout xsi:type="JsonLayout">
          <attribute name="datetime" layout="${iso8601dateformat}" />
          <attribute name="level" layout="${level:uppercase=true}" />
          <attribute name="logger" layout="${logger}" />
          <attribute name="thread" layout="${threadid}" />
          <attribute name="machine" layout="${machinename}" />
          <attribute name="message" layout="${message}" />
          <attribute name="exceptions" encode="false">
            <layout type="JsonLayout">
              <attribute name="exception" encode="false">
                <layout type="JsonLayout">
                  <attribute name="message" layout="${exception:format=message}" />
                  <attribute name="stacktrace" layout="${exception:format=stacktrace}" />
                  <attribute name="data" layout="${exception:format=data}" />
                  <attribute name="type" layout="${exception:format=type}" />
                  <attribute name="method" layout="${exception:format=method}" />
                </layout>
              </attribute>
            </layout>
          </attribute>
        </layout>
```

Which will render out something like this:

``` js
{
    "datetime": "2016-03-18T10:28:30.810",
    "level": "ERROR",
    "logger": "Fairfax.Classifieds.Ore.Company.Business.Controller.Common.ExceptionHandler",
    "thread": "82",
    "machine": "R5001764W",
    "exceptions": {
        "exception": {
            "message": "Check this out",
            "stacktrace": "   at Company.Default.Page_Load(Object sender, EventArgs e) in C:\\Source\\Company\\src\\services\\Company\\Default.aspx.cs:line 20\r\n   at System.EventHandler.Invoke(Object sender, EventArgs e)\r\n   at System.Web.UI.Control.OnLoad(EventArgs e)\r\n   at System.Web.UI.Control.LoadRecursive()\r\n   at System.Web.UI.Page.ProcessRequestMain(Boolean includeStagesBeforeAsyncPoint, Boolean includeStagesAfterAsyncPoint)",
            "type": "System.ApplicationException",
            "method": "Void Page_Load(System.Object, System.EventArgs)"
        }
    },
}
```

I would like to be able to render out is something like this:

``` js
{
    "datetime": "2016-03-18T10:28:30.810",
    "level": "ERROR",
    "logger": "Fairfax.Classifieds.Ore.Company.Business.Controller.Common.ExceptionHandler",
    "thread": "82",
    "machine": "R5001764W",
    "exceptions": {
        "exception": {
            "message": "Check this out",
            "stacktrace": "   at Company.Default.Page_Load(Object sender, EventArgs e) in C:\\Source\\Company\\src\\services\\Company\\Default.aspx.cs:line 20\r\n   at System.EventHandler.Invoke(Object sender, EventArgs e)\r\n   at System.Web.UI.Control.OnLoad(EventArgs e)\r\n   at System.Web.UI.Control.LoadRecursive()\r\n   at System.Web.UI.Page.ProcessRequestMain(Boolean includeStagesBeforeAsyncPoint, Boolean includeStagesAfterAsyncPoint)",
            "type": "System.ApplicationException",
            "method": "Void Page_Load(System.Object, System.EventArgs)"
        },
        "exception": {
            "message": "Inner Exception 1",
            "stacktrace": "   at Company.Default.Method1(Object sender, EventArgs e) in C:\\Source\\Company\\src\\services\\Company\\Default.aspx.cs:line 20\r\n   at System.EventHandler.Invoke(Object sender, EventArgs e)\r\n   at System.Web.UI.Control.OnLoad(EventArgs e)\r\n   at System.Web.UI.Control.LoadRecursive()\r\n   at System.Web.UI.Page.ProcessRequestMain(Boolean includeStagesBeforeAsyncPoint, Boolean includeStagesAfterAsyncPoint)",
            "type": "System.InvalidOperationException",
            "method": "Void Method1(System.Object, System.EventArgs)"
        }
    },
}
```

I have a technical background so I'm happy to help implement but would like some guidance on architecturally how this can be best achieved.

Many thanks,

Lach

**Comment by 304NotModified** [comment]
Cool. 

Aby idea how the config could look like? Would a double nested exception (in json) also OK?

**Comment by mccabela** [comment]
Actually Maybe a structure like this would be better?

```js
{
	"exception": {
		"message": "Check this out",
		"stacktrace": "   at Company.Default.Page_Load(Object sender, EventArgs e) in C:\\Source\\Company\\src\\services\\Company\\Default.aspx.cs:line 20\r\n   at System.EventHandler.Invoke(Object sender, EventArgs e)\r\n   at System.Web.UI.Control.OnLoad(EventArgs e)\r\n   at System.Web.UI.Control.LoadRecursive()\r\n   at System.Web.UI.Page.ProcessRequestMain(Boolean includeStagesBeforeAsyncPoint, Boolean includeStagesAfterAsyncPoint)",
		"type": "System.ApplicationException",
		"method": "Void Page_Load(System.Object, System.EventArgs)",
		"innerException": {
			"message": "Inner Exception 1",
			"stacktrace": "   at Company.Default.Method1(Object sender, EventArgs e) in C:\\Source\\Company\\src\\services\\Company\\Default.aspx.cs:line 20\r\n   at System.EventHandler.Invoke(Object sender, EventArgs e)\r\n   at System.Web.UI.Control.OnLoad(EventArgs e)\r\n   at System.Web.UI.Control.LoadRecursive()\r\n   at System.Web.UI.Page.ProcessRequestMain(Boolean includeStagesBeforeAsyncPoint, Boolean includeStagesAfterAsyncPoint)",
			"type": "System.InvalidOperationException",
			"method": "Void Method1(System.Object, System.EventArgs)",
			"innerException": {
				"message": "Inner Exception 2",
				"stacktrace": "   at Company.Default.Method1(Object sender, EventArgs e) in C:\\Source\\Company\\src\\services\\Company\\Default.aspx.cs:line 20\r\n   at System.EventHandler.Invoke(Object sender, EventArgs e)\r\n   at System.Web.UI.Control.OnLoad(EventArgs e)\r\n   at System.Web.UI.Control.LoadRecursive()\r\n   at System.Web.UI.Page.ProcessRequestMain(Boolean includeStagesBeforeAsyncPoint, Boolean includeStagesAfterAsyncPoint)",
				"type": "System.InvalidOperationException",
				"method": "Void Method1(System.Object, System.EventArgs)"
			}
		}
	}
}
```

and maybe using something like this?

```xml

          <attribute name="exception" encode="false">
            <layout type="JsonNestedExceptionLayout" source="${exception}" maxInnerExceptionLevel="5">
                  <attribute name="message" layout="${exception:format=message}" />
                  <attribute name="stacktrace" layout="${exception:format=stacktrace}" />
                  <attribute name="data" layout="${exception:format=data}" />
                  <attribute name="type" layout="${exception:format=type}" />
                  <attribute name="method" layout="${exception:format=method}" />
                  <attribute name="innerException" layout="${exception:innerexception}" />
            </layout>
          </attribute>
```

**Comment by 304NotModified** [comment]
It would be nice if we could implement this without a fully specialised layout (JsonNestedExceptionLayout isn't generic)

**Comment by mccabela** [comment]
Sounds good.  I'm only new to using this project so could you give an example of how you think it could be achieved?

**Comment by 304NotModified** [comment]
Well I don't know a good way to configure this generic.

**Comment by snakefoot** [comment]
Maybe for inspiration:

https://libraries.io/github/justeat/NLog.StructuredLogging.Json

**Comment by snakefoot** [comment]
Not the most elegant solution, but here goes:

```
        <layout type='JsonLayout'>
          <attribute name='logger' layout='${logger}' />
          <attribute name='message' layout='${message}' />
          <attribute name='exceptions' encode='false'>
            <layout type='CompoundLayout'>
                <layout type='SimpleLayout' text='{' />
                <layout type='JsonLayout' SuppressCurlyBraces='true'>
                  <attribute name='exception' encode='false'>
                    <layout type='JsonLayout'>
                      <attribute name='type' layout='${exception:format=type}' />
                      <attribute name='message' layout='${exception:format=message}' />
                    </layout>
                  </attribute>
                </layout>
                <layout type='InnerExceptions'>
                    <layout type='SimpleLayout' text=',' />
                    <layout type='JsonLayout' SuppressCurlyBraces='true'>
                      <attribute name='exception' encode='false'>
                        <layout type='JsonLayout'>
                          <attribute name='type' layout='${exception:format=type}' />
                          <attribute name='message' layout='${exception:format=message}' />
                        </layout>
                      </attribute>
                    </layout>
                </layout>
                <layout type='SimpleLayout' text='}' />
            </layout>
          </attribute>
        </layout>
```

This give the following output:

```
{ "logger": "A",
 "message": "msg",
 "exceptions": {
"exception": { "type": "System.AggregateException", "message": "One or more errors occurred." },
"exception": { "type": "System.Exception", "message": "Test exception 1" },
"exception": { "type": "System.Exception", "message": "Test Inner 1" },
"exception": { "type": "System.Exception", "message": "Test exception 2" },
"exception": { "type": "System.Exception", "message": "Test Inner 2" }
} }
```

- Have added new property **SuppressCurlyBraces** to JsonLayout.
- Have created new Layout called **InnerExceptions** (Inherits from CompoundLayout). It creates a artificial LogEventInfo that is a clone of the original LogEventInfo. It then loops through the InnerExceptions and renders the artificial LogEventInfo with each individual exception injected.

**Comment by snakefoot** [comment]
The nested inner-exception output is now supported with #2357. It just uses reflection to output all Exception-properties (so no control on the key-names)

One can override the NLog-IJsonConverter and provide a custom formatter for exception serialization.

But one might consider to have an NLog-Json-Option that specifies that all json-keys should be lowercase.

**Comment by tobz1000** [comment]
There is one problem with the current reflection-based solution - it doesn't include the name of the exception type. Currently I'm including the name of the top-level type separately, but then inner exception types are not recorded.

A generalisation of this problem would be the ability to easily add a object's type name as one of the keys in the serialised output. I could use a custom `IJsonConverter` for this, but it looks like the converter - and the `IValueFormatter` which provides it - is fixed.

What would be the best way to go about this? Either as configuration in the dependent project, or an NLog PR?

**Comment by snakefoot** [comment]
@tobz1000 I was planning to go down in the machine room after #2670 has been merged. Allowing the new XML-Layout to also benefit from the reflection logic. In that operation then I will add explicit Type-property for Exception-objects and also allow easier override of object-serialization.

**Comment by snakefoot** [comment]
@tobz1000 Created #2989 that includes Type as implicit property for Exception-objects

**Comment by snakefoot** [comment]
You can find Prerelease Nuget package here:

https://ci.appveyor.com/project/nlog/nlog/builds/20044070/artifacts

Sent from my Sony Xperia


---- Toby Dimmick wrote ----


Great! I'll give it a go when I can.

**Comment by 304NotModified** [comment]
#2989 has been merged, so NLog 4.6
